lib/iris/tests/unit/cube/test_Cube.py

```diff
@@ -1223,21 +1223,197 @@ def test__lazy(self):
+            self.assertEqual(cube.dtype, dtype)
+
+    def test_real_data_masked__mask_unset(self):
+        for dtype in self.dtypes:
+            data = ma.array([0, 1], dtype=dtype)
+            cube = Cube(data)
+            self.assertEqual(cube.dtype, dtype)
+
+    def test_real_data_masked__mask_set(self):
+        for dtype in self.dtypes:
+            data = ma.array([0, 1], dtype=dtype)
+            data[0] = ma.masked
+            cube = Cube(data)
+            self.assertEqual(cube.dtype, dtype)
+
+    def test_real_data_masked_with_dtype__mask_unset(self):
+        for dtype in self.dtypes:
+            data = ma.array([0, 1], dtype=dtype)
+            cube = Cube(data, dtype=dtype)
+            self.assertEqual(cube.dtype, dtype)
+
+    def test_real_data_masked_with_dtype__mask_set(self):
+        for dtype in self.dtypes:
+            data = ma.array([0, 1], dtype=dtype)
+            data[0] = ma.masked
+            cube = Cube(data, dtype=dtype)
+            self.assertEqual(cube.dtype, dtype)
 
-    def test_lazy(self):
-        data = np.arange(6, dtype=np.float32).reshape(2, 3)
-        cube = Cube(as_lazy_data(data))
-        self.assertEqual(cube.dtype, np.float32)
-        # Check that accessing the dtype does not trigger loading of the data.
-        self.assertTrue(cube.has_lazy_data())
+    def test_lazy_data(self):
+        for dtype in self.dtypes:
+            data = np.array([0, 1], dtype=dtype)
+            cube = Cube(as_lazy_data(data))
+            self.assertEqual(cube.dtype, dtype)
+            # Check that accessing the dtype does not trigger loading
+            # of the data.
+            self.assertTrue(cube.has_lazy_data())
+
+    def test_lazy_data_with_dtype(self):
+        for dtype in self.dtypes:
+            data = np.array([0, 1], dtype=dtype)
+            cube = Cube(as_lazy_data(data), dtype=dtype)
+            self.assertEqual(cube.dtype, dtype)
+            # Check that accessing the dtype does not trigger loading
+            # of the data.
+            self.assertTrue(cube.has_lazy_data())
+
+    def test_lazy_data_masked__mask_unset(self):
+        for dtype in self.dtypes:
+            data = ma.array([0, 1], dtype=dtype)
+            cube = Cube(as_lazy_data(data))
+            self.assertEqual(cube.dtype, dtype)
+            # Check that accessing the dtype does not trigger loading
+            # of the data.
+            self.assertTrue(cube.has_lazy_data())
+
+    def test_lazy_data_masked__mask_set(self):
+        for dtype in self.dtypes:
+            data = ma.array([0, 1], dtype=dtype)
+            data[0] = ma.masked
+            cube = Cube(as_lazy_data(data))
+            if dtype == np.dtype('float'):
+                self.assertEqual(cube.dtype, dtype)
+            else:
+                # For dask, integral masked data is converted to float
+                # to support the replacement of masks with NaNs.
+                self.assertEqual(cube.dtype, np.dtype('float'))
```

Comment:
This bit of logic may make it clearer what the expected behaviour is, but it is a bit unnecessary. Surely
```
if dtype == np.dtype('float'):
    self.assertEqual(cube.dtype, dtype)
else:
    self.assertEqual(cube.dtype, np.dtype('float'))
```
is the same as
```
self.assertEqual(cube.dtype, np.dtype('float'))
```